### ISSUE TITLE
Fully parse XMPP stream header

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2615]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2615|
|Type|defect|
|Created|2007-05-01 12:20:57Z|

The XMPP stream header parsing is fully correct nor complete. The addressing attributes `to` and `from` should cause the `thisHost` and `otherHost` attributes of the `XmlStream` be set. Also, stream IDs should only occur on streams from the receiving entity to the initiating entity.

<details><summary>Searchable metadata</summary>

```
trac-id__2615 2615
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1178022057000000 1178022057000000
changetime__1178127937000000 1178127937000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm removed owner|
|-|-|

Implemented in [branches/xmpp-stream-header-2615.](branches/xmpp-stream-header-2615.)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @ralphm|
|-|-|

 * Please add a copyright notice and a docstring to test_jabberxmlstream
 * I think the getAttribute method as a default argument (that defaults to None), so the ```rootelem.getAttribute("to") or None``` can be spelled ```rootelem.getAttribute("to")```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

I think I did the docstring and notice in another branch, but that's ok. Fixed in this branch, too.

About the `getAttribute`. I looked carefully into this when writing this code. In some places I want to have the instance variable to be `None` when the element's attribute is `''`, too.

OK for merge?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to @therve|
|-|-|

Looking at the schema for the `http://etherx.jabber.org/streams` namespace reveals that it is unclear what an empty `to` or `from` attribute would mean. Until that becomes clear, I have removed the references to `or None` as suggested.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @ralphm|
|-|-|

Hum, sorry, I should have done [#2613](https://github.com/twisted/twisted/issues/2613) first. This looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set status to `closed`|
|-|-|

(In [20161]) Properly process XMPP stream headers.

Author: ralphm
Reviewer: therve
Fixes [#2615](https://github.com/twisted/twisted/issues/2615).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
